router.spec.js

```diff
@@ -1,18 +1,141 @@
 
+process.env.NODE_ENV = 'test';
+
 const server = require('./index');
+const knex = require('./db/db');
+const user = require('./db/models/user');
+const repository = require('./db/models/repository');
+const contribution = require('./db/models/contribution');
 
 const should = chai.should();
 
 chai.use(chaiHttp);
 
+const mockRepo = {
+  id: 1,
+  owner: 1,
+  full_name: 'Bela Todo list',
+  description: 'Bela Todo list',
+  html_url: 'https://github.com/',
+  language: 'eng',
+  stargazers_count: 5
+};
+
+const mockUser = {
+  id: 1,
+  login: 'bela',
+  avatar_url: 'https://github.com/avatar',
+  html_url: 'https://github.com/avatar'
+};
+
+const mockContribution = {
+  user: 1,
+  repository: 1,
+  line_count: 50
+};
+
 describe('Web instance', () => {
-    describe('GET /hello', () => {
-      it('returns \'Hello World ! \'', async () => {
-        const response = await chai.request(server).get('/hello');
+  beforeEach(async () => {
+    await knex.raw('TRUNCATE TABLE contribution CASCADE');
+    await knex.raw('TRUNCATE TABLE repository CASCADE');
+    await knex.raw('TRUNCATE TABLE "user" CASCADE');
+    await knex.raw('ALTER SEQUENCE repository_id_seq RESTART WITH 1');
+  });
+
+  describe('GET /hello', () => {
+    it('returns \'Hello World ! \'', async () => {
+      const response = await chai.request(server).get('/hello');
+      response.should.have.status(200);
+      response.text.should.equal('Hello World !');
+    });
+  });
+
+  describe('GET /repository', async () => {
+
+    describe('/:id', () => {
+      it('returns a repository with given ID', async () => {
+        await repository.insert(mockRepo);
+        const response = await chai.request(server).get(`/repository/${mockRepo.id}`);
+        response.should.have.status(200);
+        response.body[0].should.deep.equal(mockRepo);
+      });
+    });
+
+    describe('/:id with invalid path variable', () => {
+      it('returns 403', async () => {
+        await repository.insert(mockRepo);
+        const response = await chai.request(server).get('/repository/asd');
+        response.should.have.status(403);
+      });
+    });
+
+    describe('?id=', () => {
+      it('returns a repository with given ID', async () => {
+        await repository.insert(mockRepo);
+        const response = await chai.request(server).get(`/repository?id=${mockRepo.id}`);
         response.should.have.status(200);
-        response.text.should.equal('Hello World !');
+        response.body[0].should.deep.equal(mockRepo);
       });
     });
-  }); 
+    describe('?id= with invalid query variable', () => {
+      it('returns 403', async () => {
+        const response = await chai.request(server).get('/repository?id=asd');
+        response.should.have.status(403);
+      });
+    });
+  });
+  
+  describe('POST /repository', () => {
+    it('returns the inserted object', async () => {
+      const { id, owner, full_name, description, html_url, language, stargazers_count } = mockRepo;
```

Comment:
Why destructure the entire object?

Reply:
To remove the id property. If we send a post request to this endpoint with an id included it will throw an error. ID column is incremented automatically in the DB.

Reply:
Right, I missed that one.
By the way one nice feature of postgresql is returning the inserted value, so you can get the id or any other data that is generated by the db, and you wont need to query the db to check the insertion. Have a look: https://knexjs.org/#Builder-returning